File: ui/src/api/endpoints/production.ts
/**
 * Production API Endpoints
 */
import apiClient from '../client';

export interface ProductionOrder {
  id: string;
  tenant_id: string;
  production_order_no: string;
  erp_work_order_no?: string;
  product_code: string;
  product_name?: string;
  product_rev?: string;
  line_code: string;
  target_qty: number;
  unit: string;
  started_qty: number;
  produced_qty: number;
  good_qty: number;
  defect_qty: number;
  scrap_qty: number;
  planned_start: string;
  planned_end: string;
  actual_start?: string;
  actual_end?: string;
  status: 'planned' | 'released' | 'started' | 'paused' | 'completed' | 'closed' | 'cancelled';
  current_operation?: number;
  priority: number;
  lot_no?: string;
  customer_code?: string;
  sales_order_no?: string;
  notes?: string;
  created_at: string;
  updated_at?: string;
  completion_rate?: number;
  defect_rate?: number;
}

export interface ProductionOrderListResponse {
  items: ProductionOrder[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ProductionResult {
  id: string;
  tenant_id: string;
  production_order_no: string;
  result_timestamp: string;
  shift_code: string;
  line_code: string;
  equipment_code?: string;
  operation_no: number;
  operation_name?: string;
  product_code: string;
  lot_no?: string;
  input_qty: number;
  output_qty: number;
  good_qty: number;
  defect_qty: number;
  rework_qty: number;
  scrap_qty: number;
  unit: string;
  cycle_time_sec?: number;
  takt_time_sec?: number;
  operator_code?: string;
  operator_name?: string;
  result_type: 'normal' | 'rework' | 'test' | 'trial' | 'sample';
  created_at: string;
  yield_rate?: number;
  defect_rate?: number;
}

export interface RealtimeProduction {
  id: string;
  timestamp: string;
  line_code: string;
  equipment_code: string;
  production_order_no?: string;
  product_code?: string;
  current_operation?: string;
  takt_count: number;
  good_count: number;
  defect_count: number;
  cycle_time_ms?: number;
  target_cycle_time_ms?: number;
  equipment_status?: string;
  operator_code?: string;
}

export interface DailyProductionSummary {
  date: string;
  line_code: string;
  total_input: number;
  total_output: number;
  total_good: number;
  total_defect: number;
  yield_rate: number;
  defect_rate: number;
}

export interface LineProductionStatus {
  line_code: string;
  line_name?: string;
  current_order_no?: string;
  product_code?: string;
  target_qty?: number;
  produced_qty?: number;
  completion_rate: number;
  status: string;
  equipment_count: number;
  running_count: number;
  down_count: number;
}

// API Functions

export const getWorkOrders = async (params?: {
  page?: number;
  page_size?: number;
  status?: string;
  line_code?: string;
  product_code?: string;
  start_date?: string;
  end_date?: string;
}): Promise<ProductionOrderListResponse> => {
  const response = await apiClient.get('/mes/production/work-orders', { params });
  return response.data;
};

export const getWorkOrder = async (orderId: string): Promise<ProductionOrder> => {
  const response = await apiClient.get(`/mes/production/work-orders/${orderId}`);
  return response.data;
};

export const startWorkOrder = async (orderId: string): Promise<{ message: string; order_no: string }> => {
  const response = await apiClient.post(`/mes/production/work-orders/${orderId}/start`);
  return response.data;
};

export const completeWorkOrder = async (orderId: string): Promise<{ message: string; order_no: string }> => {
  const response = await apiClient.post(`/mes/production/work-orders/${orderId}/complete`);
  return response.data;
};

export const getProductionResults = async (params?: {
  page?: number;
  page_size?: number;
  production_order_no?: string;
  line_code?: string;
  shift_code?: string;
  start_datetime?: string;
  end_datetime?: string;
}): Promise<ProductionResult[]> => {
  const response = await apiClient.get('/mes/production/results', { params });
  return response.data;
};

export const createProductionResult = async (data: Partial<ProductionResult>): Promise<ProductionResult> => {
  const response = await apiClient.post('/mes/production/results', data);
  return response.data;
};

export const getRealtimeProduction = async (params?: {
  line_code?: string;
  limit?: number;
}): Promise<RealtimeProduction[]> => {
  const response = await apiClient.get('/mes/production/results/realtime', { params });
  return response.data;
};

export const getDailyProductionAnalysis = async (params?: {
  line_code?: string;
  start_date?: string;
  end_date?: string;
}): Promise<DailyProductionSummary[]> => {
  const response = await apiClient.get('/mes/production/analysis/daily', { params });
  return response.data;
};

export const getLineProductionStatus = async (lineCode: string): Promise<LineProductionStatus> => {
  const response = await apiClient.get(`/mes/production/analysis/line/${lineCode}`);
  return response.data;
};

File: api/routers/mes/production.py
"""
Production Management API Router
"""
from datetime import datetime, date
from typing import List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select, func, and_, or_
from sqlalchemy.ext.asyncio import AsyncSession

from api.database import get_db
from api.config import settings
from api.models.mes.production import ProductionOrder, ProductionResult, RealtimeProduction
from api.schemas.mes.production import (
    ProductionOrderCreate,
    ProductionOrderUpdate,
    ProductionOrderResponse,
    ProductionOrderListResponse,
    ProductionOrderStatus,
    ProductionResultCreate,
    ProductionResultResponse,
    RealtimeProductionResponse,
    DailyProductionSummary,
    LineProductionStatus,
)
from api.services.mock_data import MockDataService


router = APIRouter(prefix="/production", tags=["MES - Production"])


# ==================== Production Orders ====================

@router.get("/work-orders", response_model=ProductionOrderListResponse)
async def get_work_orders(
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    status: Optional[ProductionOrderStatus] = None,
    line_code: Optional[str] = None,
    product_code: Optional[str] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
):
    """Get list of production orders (work orders)"""
    tenant_id = UUID(settings.default_tenant_id)

    # Build query
    query = select(ProductionOrder).where(ProductionOrder.tenant_id == tenant_id)

    if status:
        query = query.where(ProductionOrder.status == status.value)
    if line_code:
        query = query.where(ProductionOrder.line_code == line_code)
    if product_code:
        query = query.where(ProductionOrder.product_code == product_code)
    if start_date:
        query = query.where(ProductionOrder.planned_start >= datetime.combine(start_date, datetime.min.time()))
    if end_date:
        query = query.where(ProductionOrder.planned_start <= datetime.combine(end_date, datetime.max.time()))

    # Count total
    count_query = select(func.count()).select_from(query.subquery())
    total = await db.scalar(count_query) or 0

    # Paginate
    query = query.order_by(ProductionOrder.planned_start.desc())
    query = query.offset((page - 1) * page_size).limit(page_size)

    result = await db.execute(query)
    orders = result.scalars().all()

    # Calculate completion and defect rates
    items = []
    for order in orders:
        order_dict = ProductionOrderResponse.model_validate(order).model_dump()
        if order.target_qty and order.target_qty > 0:
            order_dict["completion_rate"] = float(order.produced_qty / order.target_qty) * 100
        if order.produced_qty and order.produced_qty > 0:
            order_dict["defect_rate"] = float(order.defect_qty / order.produced_qty) * 100
        items.append(ProductionOrderResponse(**order_dict))

    return ProductionOrderListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=(total + page_size - 1) // page_size,
    )


@router.get("/work-orders/{order_id}", response_model=ProductionOrderResponse)
async def get_work_order(
    order_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Get production order by ID"""
    tenant_id = UUID(settings.default_tenant_id)

    query = select(ProductionOrder).where(
        and_(
            ProductionOrder.id == order_id,
            ProductionOrder.tenant_id == tenant_id,
        )
    )
    result = await db.execute(query)
    order = result.scalar_one_or_none()

    if not order:
        raise HTTPException(status_code=404, detail="Production order not found")

    return ProductionOrderResponse.model_validate(order)


@router.post("/work-orders/{order_id}/start")
async def start_work_order(
    order_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Start a production order"""
    tenant_id = UUID(settings.default_tenant_id)

    query = select(ProductionOrder).where(
        and_(
            ProductionOrder.id == order_id,
            ProductionOrder.tenant_id == tenant_id,
        )
    )
    result = await db.execute(query)
    order = result.scalar_one_or_none()

    if not order:
        raise HTTPException(status_code=404, detail="Production order not found")

    if order.status not in ["planned", "released"]:
        raise HTTPException(status_code=400, detail=f"Cannot start order with status: {order.status}")

    order.status = "started"
    order.actual_start = datetime.utcnow()
    await db.commit()

    return {"message": "Production order started", "order_no": order.production_order_no}


@router.post("/work-orders/{order_id}/complete")
async def complete_work_order(
    order_id: UUID,
    db: AsyncSession = Depends(get_db),
):
    """Complete a production order"""
    tenant_id = UUID(settings.default_tenant_id)

    query = select(ProductionOrder).where(
        and_(
            ProductionOrder.id == order_id,
            ProductionOrder.tenant_id == tenant_id,
        )
    )
    result = await db.execute(query)
    order = result.scalar_one_or_none()

    if not order:
        raise HTTPException(status_code=404, detail="Production order not found")

    if order.status != "started":
        raise HTTPException(status_code=400, detail=f"Cannot complete order with status: {order.status}")

    order.status = "completed"
    order.actual_end = datetime.utcnow()
    await db.commit()

    return {"message": "Production order completed", "order_no": order.production_order_no}


# ==================== Production Results ====================

@router.get("/results", response_model=List[ProductionResultResponse])
async def get_production_results(
    db: AsyncSession = Depends(get_db),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    production_order_no: Optional[str] = None,
    line_code: Optional[str] = None,
    shift_code: Optional[str] = None,
    start_datetime: Optional[datetime] = None,
    end_datetime: Optional[datetime] = None,
):
    """Get production results"""
    tenant_id = UUID(settings.default_tenant_id)

    query = select(ProductionResult).where(ProductionResult.tenant_id == tenant_id)

    if production_order_no:
        query = query.where(ProductionResult.production_order_no == production_order_no)
    if line_code:
        query = query.where(ProductionResult.line_code == line_code)
    if shift_code:
        query = query.where(ProductionResult.shift_code == shift_code)
    if start_datetime:
        query = query.where(ProductionResult.result_timestamp >= start_datetime)
    if end_datetime:
        query = query.where(ProductionResult.result_timestamp <= end_datetime)

    query = query.order_by(ProductionResult.result_timestamp.desc())
    query = query.offset((page - 1) * page_size).limit(page_size)

    result = await db.execute(query)
    results = result.scalars().all()

    return [ProductionResultResponse.model_validate(r) for r in results]


@router.post("/results", response_model=ProductionResultResponse)
async def create_production_result(
    data: ProductionResultCreate,
    db: AsyncSession = Depends(get_db),
):
    """Create new production result"""
    tenant_id = UUID(settings.default_tenant_id)

    result = ProductionResult(
        tenant_id=tenant_id,
        result_timestamp=data.result_timestamp or datetime.utcnow(),
        **data.model_dump(exclude={"result_timestamp"})
    )
    db.add(result)
    await db.commit()
    await db.refresh(result)

    return ProductionResultResponse.model_validate(result)


@router.get("/results/realtime")
async def get_realtime_production(
    db: AsyncSession = Depends(get_db),
    line_code: Optional[str] = None,
    limit: int = Query(100, ge=1, le=500),
):
    """Get realtime production data (with mock data fallback)

    Returns data structure expected by Dashboard:
    {
        "hourly": [...],
        "lines": [...],
        "summary": {...}
    }
    """
    try:
        tenant_id = UUID(settings.default_tenant_id)

        query = select(RealtimeProduction).where(RealtimeProduction.tenant_id == tenant_id)

        if line_code:
            query = query.where(RealtimeProduction.line_code == line_code)

        query = query.order_by(RealtimeProduction.timestamp.desc()).limit(limit)

        result = await db.execute(query)
        data = result.scalars().all()

        if not data:
            raise Exception("No realtime data found, using mock data")

        return [RealtimeProductionResponse.model_validate(d) for d in data]
    except Exception:
        # Return mock data formatted for Dashboard
        return MockDataService.get_realtime_production_formatted(line_code)


# ==================== Production Analysis ====================

@router.get("/analysis/daily")
async def get_daily_production_analysis(
    db: AsyncSession = Depends(get_db),
    line_code: Optional[str] = None,
    start_date: Optional[date] = None,
    end_date: Optional[date] = None,
):
    """Get daily production summary (with mock data fallback)

    Returns data structure expected by Dashboard:
    {
        "daily_data": [{today's data}],
        "product_distribution": [...],
        "by_line": [...],
        ...
    }
    """
    try:
        tenant_id = UUID(settings.default_tenant_id)

        # Build aggregation query
        query = select(
            func.date_trunc('day', ProductionResult.result_timestamp).label('date'),
            ProductionResult.line_code,
            func.sum(ProductionResult.input_qty).label('total_input'),
            func.sum(ProductionResult.output_qty).label('total_output'),
            func.sum(ProductionResult.good_qty).label('total_good'),
            func.sum(ProductionResult.defect_qty).label('total_defect'),
        ).where(ProductionResult.tenant_id == tenant_id)

        if line_code:
            query = query.where(ProductionResult.line_code == line_code)
        if start_date:
            query = query.where(ProductionResult.result_timestamp >= datetime.combine(start_date, datetime.min.time()))
        if end_date:
            query = query.where(ProductionResult.result_timestamp <= datetime.combine(end_date, datetime.max.time()))

        query = query.group_by(
            func.date_trunc('day', ProductionResult.result_timestamp),
            ProductionResult.line_code,
        ).order_by(func.date_trunc('day', ProductionResult.result_timestamp).desc())

        result = await db.execute(query)
        rows = result.all()

        if not rows:
            raise Exception("No daily analysis data found, using mock data")

        summaries = []
        for row in rows:
            total_input = float(row.total_input or 0)
            total_good = float(row.total_good or 0)
            total_defect = float(row.total_defect or 0)

            yield_rate = (total_good / total_input * 100) if total_input > 0 else 0
            defect_rate = (total_defect / total_input * 100) if total_input > 0 else 0

            summaries.append(DailyProductionSummary(
                date=row.date,
                line_code=row.line_code,
                total_input=row.total_input or 0,
                total_output=row.total_output or 0,
                total_good=row.total_good or 0,
                total_defect=row.total_defect or 0,
                yield_rate=yield_rate,
                defect_rate=defect_rate,
            ))

        return summaries
    except Exception:
        # Return mock data formatted for Dashboard
        return MockDataService.get_daily_production_analysis_formatted()


@router.get("/analysis/line/{line_code}", response_model=LineProductionStatus)
async def get_line_production_status(
    line_code: str,
    db: AsyncSession = Depends(get_db),
):
    """Get production status for a specific line"""
    tenant_id = UUID(settings.default_tenant_id)

    # Get current active order
    order_query = select(ProductionOrder).where(
        and_(
            ProductionOrder.tenant_id == tenant_id,
            ProductionOrder.line_code == line_code,
            ProductionOrder.status == "started",
        )
    ).order_by(ProductionOrder.actual_start.desc()).limit(1)

    result = await db.execute(order_query)
    current_order = result.scalar_one_or_none()

    # Get equipment count (simplified - would need equipment table join)
    equipment_count = 10  # Placeholder
    running_count = 8
    down_count = 2

    completion_rate = 0.0
    if current_order and current_order.target_qty and current_order.target_qty > 0:
        completion_rate = float(current_order.produced_qty / current_order.target_qty) * 100

    return LineProductionStatus(
        line_code=line_code,
        line_name=f"Line {line_code}",
        current_order_no=current_order.production_order_no if current_order else None,
        product_code=current_order.product_code if current_order else None,
        target_qty=current_order.target_qty if current_order else None,
        produced_qty=current_order.produced_qty if current_order else None,
        completion_rate=completion_rate,
        status="running" if current_order else "idle",
        equipment_count=equipment_count,
        running_count=running_count,
        down_count=down_count,
    )
